### ISSUE TITLE
expo cli crash while rebuilding web package

### ISSUE THREAD

**Issue by pihomeserver** [expo/expo-cli#1256]
## Description

Using the expo-cli to work a project, tested as a web package
The build works the first time, the live refresh crashes the client and it should be restarted manually

## Expected Behavior

Previous version had live refresh full operational

## Observed Behavior

Here is the stack trace of the first run after the build (tunnel is up and web page opens successfully) : 

```
To run the app with live reloading, choose one of:
  • Sign in as @pihomeserver in Expo Client on Android or iOS. Your projects will automatically appear in the "Projects" tab.
  • Scan the QR code above with the Expo app (Android) or the Camera app (iOS).
  • Press a for Android emulator, or i for iOS simulator.
  • Press e to send a link to your phone with email.
┌──────────────────────────────────────────────────────────────────────────┐
│                                                                          │
│   Web support in Expo is experimental and subject to breaking changes.   │
│   Do not use this in production yet.                                     │
│                                                                          │
└──────────────────────────────────────────────────────────────────────────┘

 React  You can now view AuthTest in the browser.

  Local:            http://localhost:19006/
  On Your Network:  http://192.168.1.10:19006/

Note that the development build is not optimized.

 › To create a production build, run expo build:web
 › Press Ctrl+C to exit.

 Expo  Press ? to show a list of all available commands.
```

Then if i change a source file : 
```
Compiling...

  build [========            ] 40%TypeError: compilation.getPathWithInfo is not a function
    at /Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/@expo/webpack-config/node_modules/webpack/lib/HotModuleReplacementPlugin.js:287:26
    at SyncHook.eval (eval at create (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:14:1)
    at SyncHook.lazyCompileHook (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/Hook.js:154:20)
    at /Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/webpack/lib/Compilation.js:1361:37
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:7:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.seal (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/webpack/lib/Compilation.js:1304:27)
    at /Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/webpack/lib/Compiler.js:665:18
    at /Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/webpack/lib/Compilation.js:1224:4
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:18:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.finish (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/webpack/lib/Compilation.js:1216:28)
    at /Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/webpack/lib/Compiler.js:662:17
    at _done (eval at create (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at eval (eval at create (/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:32:22)
    at /Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/react-dev-utils/ModuleNotFoundPlugin.js:114:15
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ start: `expo start`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/chris/.npm/_logs/2019-11-22T16_32_23_648Z-debug.log
```

## Environment

```
  Expo CLI 3.8.0 environment info:
    System:
      OS: macOS 10.15.1
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 13.0.1 - ~/.nvm/versions/node/v13.0.1/bin/node
      Yarn: 1.19.1 - ~/.yarn/bin/yarn
      npm: 6.13.1 - ~/.nvm/versions/node/v13.0.1/bin/npm
    IDEs:
      Xcode: 11.2.1/11B500 - /usr/bin/xcodebuild
    npmPackages:
      @types/react: ^16.8.23 => 16.9.11
      @types/react-native: ^0.57.65 => 0.57.65
      expo: ^35.0.0 => 35.0.1
      react: 16.8.3 => 16.8.3
      react-native: https://github.com/expo/react-native/archive/sdk-35.0.0.tar.gz => 0.59.8
      react-navigation: ^4.0.10 => 4.0.10
    npmGlobalPackages:
      expo-cli: 3.8.0
```

Web browser is Brave Version 1.0.0 Chromium: 78.0.3904.97 (Build officiel) (64 bits)

## Reproducible Demo

```
expo init test
```
<select 2nd option>
```
cd test
yarn start
```
type `w` to start the web version
Once started replace in `App.tsx` : 
```
backgroundColor: '#fff'
```
with 
```
backgroundColor: '#ff0'
```
Same project opened on an iPhone updates successfully so seems related only to live module reload for the web version

**Comment by Smarticles101** [comment]
Can confirm; I have having the same problem. Webpack works on the first run, but once you change a source file it errors out with `TypeError: compilation.getPathWithInfo is not a function`

**Comment by pihomeserver** [comment]
Seems to have to downgrade webpack version : https://github.com/gatsbyjs/gatsby/issues/19081#issuecomment-548711915

I'll give a try to that

**Comment by voland62** [comment]
Same problem.
  I can't figure out how can we downgrade webpack, becouse it is indirect dependency...

**Comment by pihomeserver** [comment]
Maybe with a downgrade of the expo cli ? Not sure also how to do that

**Comment by voland62** [comment]
> Maybe with a downgrade of the expo cli ? Not sure also how to do that

I tryed. Didn't help for me...
I guess they both(3.5.0 & 3.9.0) have unbounded version dependency something like webpack_version >  some_errata_version ...

**Comment by Arlen22** [comment]
Wish there was a way in NPM to keep track of what versions changed when in a dependency tree then we could roll it back to a specific date

**Comment by Arlen22** [comment]
> I can't figure out how can we downgrade webpack, becouse it is indirect dependency...

Since the problem started two days ago apparently, you could just install the webpack version (or whatever the culprit was) manually and it should still be in the range of compatible versions since expo itself wasn't updated.

**Comment by pihomeserver** [comment]
Installed globally webpack 4.41.2 manually (released mid october) and same issue

**Comment by voland62** [comment]
> Installed globally webpack 4.41.2 manually (released mid october) and same issue

Looking at this:
`/Users/chris/.nvm/versions/node/v13.0.1/lib/node_modules/expo-cli/node_modules/@expo/webpack-config/node_modules/webpack/lib/HotModuleReplacementPlugin.js:287:26`
It uses some internal installed webpack, not the global one...

**Comment by Arlen22** [comment]
You would need to manually delete that webpack folder and drop in a copy of one you installed globally. Remember, on of webpack's dependancies may be the culprit.

**Comment by Arlen22** [comment]
I found the current definition for the function (I think), and modified it to this, which does not throw, nor is the counter accessible when I break just before the error. 

```js
        counter = 0;
	/**
	 * @param {string} filename used to get asset path with hash
	 * @param {TODO=} data // TODO: figure out this param type
	 * @returns {{ path: string, info: AssetInfo }} interpolated path and asset info
	 */
	get getPathWithInfo() { console.log(counter++); return (filename, data) => {
		data = data || {};
		data.hash = data.hash || this.hash;
		return this.mainTemplate.getAssetPathWithInfo(filename, data);
	}; }
	set getPathWithInfo(v) {
		throw new Error("Tried to set getPathWithInfo");
	}
```

**Comment by Arlen22** [comment]
@brentvatne @EvanBacon Please see my last comment, which is heavily edited. I believe I have identified the culprit.

**Comment by Arlen22** [comment]
I think this should give us some clues. I don't really understand yarn.lock, but it gives us the best day to day snapshot I think.

https://github.com/expo/expo-cli/commit/9cac56506c22be6676d5742b1180a6665dbd26fb#diff-8ee2343978836a779dc9f8d6b794c3b2

**Comment by Arlen22** [comment]
Ok, that didn't give me much, but I ran `npm list` and discovered that there are two specified. The first is the one I already mentioned under @expo/webpack-config, the other is under @expo/xdl, which lists the old version. I tried manually running `npm install webpack@4.41.2` in the expo-cli global install, but this probably just moved `webpack@4.39.0` into the `@expo/xdl` directory or something, I don't know. It didn't make a difference. 

So the culprit is `@expo/xdl` conflicting with `@expo/webpack-config`. I don't know which is correct or anything like that.

**Comment by Arlen22** [comment]
It does indeed move it to the `@expo/xdl/node_modules` folder. So I ran it again with the `--prod` flag (pretty important), and then renamed the webpack in `@expo/xdl/node_modules` to something different so it couldn't find it. Now it compiles and reloads the web page properly, but it gives me this warning: 

> Conflict: Multiple assets emit different content to the same filename a6832bfc2a0e11892e42.map

**Comment by Arlen22** [comment]
For anyone who's curious, I decided to go back to the default install until it's fixed in case there's something else I'm not aware of, because the error that resulted instead is rather dubious. Two assets with different content emitting to the same hash-named file is not a hash collision three times in a row.

**Comment by Arlen22** [comment]
I'm not seeing a shrinkwrap file, but there is a yarn.lock file which accomplishes the same purpose since NPM presumably doesn't block yarn.lock files. So going back in the history, we should try 3.8.1, 3.8.0, and 3.7.1. The last one should work, I'm assuming, because I think 3.8.0 was when the change happened.

**Comment by brentvatne** [comment]
@EvanBacon ping

**Comment by Arlen22** [comment]
> I'm not seeing a shrinkwrap file, but there is a yarn.lock file which accomplishes the same purpose since NPM presumably doesn't block yarn.lock files. So going back in the history, we should try 3.8.1, 3.8.0, and 3.7.1. The last one should work, I'm assuming, because I think 3.8.0 was when the change happened.

Nevermind, that yarn.lock is for development. I wasn't thinking.

**Comment by Arlen22** [comment]
Manually installing 3.7.1 then changing the semver in @expo/xdl dependency on @expo/webpack-config to the exact version specify instead of `^` fixed the problem for the moment.

**Comment by ludomout** [comment]
👋 same problem
Adding `  "resolutions": {
    "expo-cli/**/webpack": "4.39.0"
  },` to `package.json` seems to resolve it

**Comment by luisrivas** [comment]
> 👋 same problem
> Adding ` "resolutions": { "expo-cli/**/webpack": "4.39.0" },` to `package.json` seems to resolve it

Thanks @ludomout , It's not maybe a definitive solution, but just saved my day

**Comment by marcioferlan** [comment]
Where did you guys add this line?

> 👋 same problem
> Adding ` "resolutions": { "expo-cli/**/webpack": "4.39.0" },` to `package.json` seems to resolve it

I removed my `node_modules` folder, added that line to the my app's `package.json` file and ran `yarn install`, but I'm still getting the same error.

**Comment by Arlen22** [comment]
You might need to use npm install —prod?

**Comment by marcioferlan** [comment]
I managed to solve it by installing `webpack@4.39.0` into my global `expo-cli` installation:

These were the steps to fix it:

```
cd /Users/marcio/.nvm/versions/node/v8.15.1/lib/node_modules/expo-cli
cd node_modules/@expo/webpack-config
npm i webpack@4.39.0
```

Then when I started my app with `expo start` it reloaded correctly after file changes. 🎉

**Comment by Arlen22** [comment]
You should use the --prod flag first, because global packages are installed as production packages, and I did notice a difference in my initial testing. If it doesn't work then try without the --prod flag.

**Comment by Arlen22** [comment]
> ```
> cd /Users/marcio/.nvm/versions/node/v8.15.1/lib/node_modules/expo-cli
> cd node_modules/@expo/webpack-config
> npm i webpack@4.39.0
> ```

This solved it for me, but the "resolutions" property in package.json did not. Saving an npm shrinkwrap :)

**Comment by Arlen22** [comment]
This solution has one side-effect. After several refreshes, the sourcemap begins to drift and errors show up in the wrong place.

**Comment by luisrivas** [comment]
> Where did you guys add this line?
> 
> > 👋 same problem
> > Adding ` "resolutions": { "expo-cli/**/webpack": "4.39.0" },` to `package.json` seems to resolve it
> 
> I removed my `node_modules` folder, added that line to the my app's `package.json` file and ran `yarn install`, but I'm still getting the same error.

In order to use that solution, you need to install *temporally* `expo-cli` in your project a run this local version using a npm script, for example, "web": "expo start --web "

**Comment by pihomeserver** [comment]
Thank you all for tips and the solution !

**Comment by rizkyramadhan** [comment]
@pihomeserver how do you update exo client ? is it expo-cli ?

**Comment by pihomeserver** [comment]
@rizkyramadhan yes that's it. 
I made an update of expo-cli via yarn, deleted node_modules and rerun yarn install

**Comment by Lidiya777** [comment]
> Same problem.
> I can't figure out how can we downgrade webpack, becouse it is indirect dependency...

Андрей, привет! напиши мне на kenzo@ukr.net

**Comment by Vn-ChemGio** [comment]
same problem, with expo 37.0.0

**Comment by Vn-ChemGio** [comment]
I'm using expo 3.20.1 had that problem :(

**Comment by marivaaldo** [comment]
I solved after defined webpack version to 4.39.0 in my devDependencies on package.json.

Like this:
```json
{
     "devDependencies": {
          "webpack": "4.39.0",
          ...
     },
     "dependencies": {
          ...
     },
     ...
}
```
